Modules/posixmodule.c

```diff
@@ -5553,6 +5520,98 @@ os__path_normpath_impl(PyObject *module, path_t *path)
     return result;
 }
 
+/*[clinic input]
+os._path_abspath
+
+    path: path_t(make_wide=True, nonstrict=True)
+    /
+
+Return an absolute path.
+[clinic start generated code]*/
+
+static PyObject *
+os__path_abspath_impl(PyObject *module, path_t *path)
+/*[clinic end generated code: output=bb40fbf3be7251a4 input=8870bbcaa7c920a1]*/
+{
+    Py_ssize_t abs_len;
+    wchar_t *abs, *abs_buf = NULL;
+
+    wchar_t *path_buf = (wchar_t *)path->wide;
+    Py_ssize_t path_len = path->length;
+    int use_bytes = PyBytes_Check(path->object);
+#ifdef MS_WINDOWS
+    if (wcslen(path_buf) != path_len) {
+        return PyErr_Format(PyExc_ValueError,
+                            "_path_abspath: embedded null character in path");
+    }
+    // Preserve `.\` for qualified referencing
+    abs = _Py_normpath_and_size(path_buf, path_len, 0, &abs_len, 1);
+    if (abs_len == 0 || (abs_len == 1 && abs[0] == L'.')) {
+        return posix_getcwd(use_bytes);
+    }
+    if (_PyOS_getfullpathname(abs, &abs_buf) < 0) {
+        return win32_error_object("GetFullPathNameW", path->object);
+    }
+    abs = abs_buf;
+    abs_len = wcslen(abs_buf);
+#else
+    if (path_len == 0 || (path_len == 1 && path_buf[0] == L'.')) {
+        return posix_getcwd(use_bytes);
+    }
+
+    if (_Py_isabs(path_buf)) {
+        abs = _Py_normpath_and_size(path_buf, path_len, 0, &abs_len, 0);
+    }
+    else {
+        PyObject *cwd_obj = posix_getcwd(0);
+        if (!cwd_obj) {
+            return NULL;
+        }
+        Py_ssize_t cwd_len;
+        wchar_t *cwd_buf = PyUnicode_AsWideCharString(cwd_obj, &cwd_len);
+        Py_DECREF(cwd_obj);
+        if (!cwd_buf) {
+            return NULL;
+        }
+
+        int add_sep = cwd_buf[cwd_len - 1] != SEP;
+        Py_ssize_t prefix_len = cwd_len + add_sep;
+        abs_len = prefix_len + path_len;
+
+        if ((size_t)abs_len + 1 > (size_t)PY_SSIZE_T_MAX / sizeof(wchar_t)) {
+            PyMem_Free(cwd_buf);
+            PyErr_SetString(PyExc_OverflowError, "path is too long");
+            return NULL;
+        }
+
+        abs_buf = PyMem_RawMalloc(((size_t)abs_len + 1) * sizeof(wchar_t));
+        if (!abs_buf) {
+            PyMem_Free(cwd_buf);
```

Comment:
This repeated code and fine-grained resource management is begging for code changes that introduce memory leaks. This is an uncommon patter in "Modules/posixmodule.c", which makes extensive use of `goto exit` and `goto fail` to track resource management in a way that reduces repeated code and minimizes the chance for introducing bugs. This is all just to avoid declaring a pointer at function scope that only gets used in a branch? I'll try to write another version of the function in hopes that it gets accepted.

Reply:
Yes, it is a symptom of code that needs refactoring. We (should) try really hard to keep the code clean, readable, and maintainable.